book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,437 @@
+    &mut self,
+    prepared: PreparedBlock,
+    finalized_tip_history_tree: HistoryTree,
+    finalized_value_pool: ValueBalance<NonNegative>,
+) -> Result<(), ValidateContextError> {
+    ..
+    check_value_pools(finalized_value_pools, prepared.value_balance)?;
+    ..
+}
+
+fn check_value_pools(
+    finalized_value_pools: ValueBalance<NonNegative>,
+    block_value_balance: ValueBalance<NegativeAllowed>
+) -> Result<()> {
+    if finalized_value_pools + block_value_balance < 0 {
+        Err("Value pool can't be negative");
+    }
+    Ok(())
+}
+```
+
+##### Consensus rule
+
+The chain value pool balance rules apply to Block transactions, but they are optional for Mempool transactions:
+> Nodes MAY relay transactions even if one or more of them cannot be mined due to the aforementioned restriction.
+
+https://zips.z.cash/zip-0209#specification
+
+Since Zebra does chain value pool balance validation in the state, we want to skip verifying the speculative chain balance of Mempool transactions.
+
+### Changes to finalized state
+
+The state service is what will call `commit_block()` and `commit_new_chain()`. We need to pass the value pool from the disk into this functions.
+
+```
+self.mem
+    .commit_block(prepared, self.disk.history_tree(), self.disk.get_pool())?;
+..
+self.mem
+    .commit_new_chain(prepared, self.disk.history_tree(), self.disk.get_pool())?;
+```
+
+We now detail what is needed in order to have the `get_pool()` method available.
+
+#### Serialization of `ValueBalance<C>`
+
+In order to save `ValueBalance` into the disk database we must implement `IntoDisk` and `FromDisk` for `ValueBalance` and for `Amount`:
+
+```rust
+
+impl IntoDisk for ValueBalance<C> {
+    ..
+}
+
+impl FromDisk for ValueBalance<C> {
+    ..
+}
+
+impl IntoDisk for Amount {
+    type Bytes = [u8; 8];
+
+    fn as_bytes(&self) -> Self::Bytes {
+        self.to_bytes()
+    }
+}
+
+impl FromDisk for Amount {
+    fn from_bytes(bytes: impl AsRef<[u8]>) -> Self {
+        let array = bytes.as_ref().try_into().unwrap();
+        i64::from_be_bytes(array).try_into().unwrap()
+    }
+}
+```
+
+#### Changes to `zebra-state/src/service/finalized_state.rs`
+
+First we add a column for all the pools: transparent, sprout, sapling, orchard:
+
+```rust
+rocksdb::ColumnFamilyDescriptor::new("value_pools", db_options.clone()),
+```
+
+At block commit(`commit_finalized_direct()`) we create the handles for the new columns:
+
+```rust
+let value_pools = self.db.cf_handle("value_pools").unwrap();
+```
+
+Next we save each value pool into the field for each upcoming block except for the genesis block:
+
+```rust
+// Consensus rule: The block height of the genesis block is 0
+// https://zips.z.cash/protocol/protocol.pdf#blockchain
+if height == block::Height(0) {
+    batch.zs_insert(value_pools, height, ValueBalance::default());
+} else {
+    let current_pool = self.get_pool();
+    batch.zs_insert(value_pools, height, (current_pool + block.value_pool())?);
```

Comment:
We already calculated the value balances and stored them in the `PreparedBlock`.
Do you want to copy that field to the `FinalizedBlock`?

This saves a small amount of CPU time when we finalize a block with lots of transactions.

I don't mind either way.

```suggestion
    batch.zs_insert(value_pools, height, (current_pool + finalized.block_value_balance)?);
```

Reply:
Implemented this at https://github.com/ZcashFoundation/zebra/pull/2430/commits/ee5dc13d110872908002c6d9d67bef48c4217020